Randomly Answer Question Fixes

BryceStevenWilley opened the pull request:
Trying to get randomly answer questions working on a real interview, but it's a bit brittle:

*  would occasionally run into null errors. To fix, just added more null checks in [this commit](https://github.com/SuffolkLITLab/ALKiln/commit/1266101d202f4a11518d7d43deb6e1cbd95a6191)
* on the Affidavit of Indigency, the kickout screen has an exit button that takes you to courtformsonline. Unfortunately, Kiln will take you there, but not realize you've left the interview, and thinks it's hit an error page. So we now navigate back to the interview if we leave it somehow.
* Fix #628, only presses the back button 10% of the time, not 50% of the time. The latter made it very improbable that you'd make it very far into the interview.
* Addresses some issues with radio buttons: only the last one would ever be pressed, which always led you to a kickout screen for this interview. Improved it a little bit, but still favors the later buttons due to implementation.

Some existing bugs that are still around:
* each radio button is considered separately, heavily weighting the answers towards the latter radio buttons; the last button has a 50% chance of being clicked, and the 2nd to last has a 25% chance, etc. https://github.com/SuffolkLITLab/ALKiln/issues/637
* doesn't work well with list collect screens, it tries to answer every single list item that it sees on screen (~20 or 30 I think?) and times out. https://github.com/SuffolkLITLab/ALKiln/issues/638
*  still running into weird bugs on certain screens when Kiln thinks there's no content. Unclear what page it's actually trying to query (maybe a not-yet loaded page?) https://github.com/SuffolkLITLab/ALKiln/issues/809

Even with those bugs, this PR should continue, definitely improves the situation a bit, though I wouldn't routinely run it for an interview yet, as there are still a lot of false "found an error screen"s.

BryceStevenWilley committed:
Fix errors in random tests

BryceStevenWilley committed:
Navigate back to the interview if you randomly leave

Happens with Exit buttons in kickout screens, which
randomly answer questions will get to often.

BryceStevenWilley committed:
Press the back button less often in random runs

Reduces the chance a back button is pressed to max 10%.

Fixes #628.

BryceStevenWilley committed:
Press other radio buttons than the last

Several bugs when pressing radio buttons:
* radio buttons were mistakenly found to be custom elements? Not sure why, but
  I did have to hard code an exception there.
* radio buttons were always being pressed, regardless of the actual desired
  answer

Still present, and not desirable: each radio button is considered separately,
heavily weighting the answers towards the latter radio buttons; the last button
has a 50% chance of being clicked, and the 2nd to last has a 25% chance, etc.

BryceStevenWilley committed:
Maybe avoid null error

plocket committed:
Fix #629, artifact folders not being saved in GitHub

Also adjust internal test names to be a bit shorter and more
flexible in case we want to make them more modular in future. Not
as sure about that one.

plocket committed:
Bump minor v to fix artifacts not being found

plocket commented on lib/scope.js:
```diff
@@ -1742,7 +1742,13 @@ module.exports = {
 
-      let type = await handle.evaluate( elem => { return elem.getAttribute('type'); });
+      let type = await handle.evaluate( elem => { 
+        if (elem) {
+          return elem.getAttribute('type'); 
+        } else {
+          return '';
+        }
```
If the handle doesn't actually represent an element, should this function stop completely? Should it throw an error? Maybe I've forgotten something else this does. I'll try to take a closer look, but may not get to it.

BryceStevenWilley replied:
I don't know what it should do. I had to add this fix because `elem` was `null`, just to avoid running into uncertain null errors. Since we don't know the context of the error fully, it's likely something with puppeteer (or our use of puppeteer), and I'd rather just not have that happen. Throwing another error IMO would be the same issue there, we'd have to catch the error everywhere to else to prevent errors propagating.

I'd be willing to add a log, though it can't go right there (because that's JS that's executing in the browser, not this machine).

plocket replied:
Did a little experiment in the browser [and a bit of research]. When a type is null, the default type is 'text', so we'll use that instead of an empty string. I can do that when I'm working on peeling out the radio button stuff. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#input_types

plocket commented on lib/scope.js:
```diff
@@ -2741,9 +2746,8 @@ module.exports = {
       // If this is a custom datatype
       let custom_datatype = await scope.get_custom_datatype( scope, { handle });
-      let is_custom_datatype = scope.get_random_input_for[ custom_datatype ] !== undefined;
+      let is_custom_datatype = scope.get_random_input_for[ custom_datatype ] !== undefined && type != 'radio';
       if ( is_custom_datatype ) {
```
```suggestion
      let is_custom_datatype = scope.get_random_input_for[ custom_datatype ] !== undefined;
      if ( is_custom_datatype && type !== `radio` ) {
```

Otherwise, the name `is_custom_datatype` doesn't mean what you'd expect it to mean. I'm not sure why this excludes custom datatypes that are radio buttons. Where do we deal with custom datatypes that are radio buttons?

BryceStevenWilley replied:
First, I don't think we have any custom datatypes that are radio buttons.

Second, this probably shouldn't need to exist, I just know that `get_custom_datatype` and `is_custom_datatype` would say that radio buttons were custom datatypes when they weren't actually. I would guess that `get_custom_datatype` isn't correct, but I don't know how to fix it without breaking other things.

plocket replied:
Made #807

plocket commented on lib/scope.js:
```diff
@@ -1759,8 +1759,13 @@ module.exports = {
-        if ( type === `radio` ) { await label.evaluate( elem => { return elem.click(); }); }
-        else { await scope.setCheckbox( scope, { label, answer }); }
+        if ( type === `radio` ) {
+          if (answer.toLowerCase() == "true") {
```
After a bit of trial and error, I have a guess as to what happened here. In the non-random flow, if a radio button is going to be pressed, only that radio button will get into here. The `answer` is the DOM `value` of the radio button. A an `answer` for a 'no' radio button will probably be "False". If it's the set of choices, it'll be something like "radio_other_opt_3".

That is, by the time the handle is in here, this radio button should definitely be pressed, no matter what the answer is.

If we want to control the flow of random radio buttons, I think that needs to happen upstream where random answers are dealt with. If we don't want a radio button to be pressed, it shouldn't get sent in here.

BryceStevenWilley replied:
Hm, I get it. A weird place for a parameter to have no effect, but I don't know if there's a refactor that makes it cleaner.

From what I can tell, there's no existing function to get the elements of a radio button? i.e. which values it can be set to? I'll have to let this sit until I can figure that out.

plocket replied:
There's no way to get the options of a radio button group right now. Maybe separate the radio button stuff from the other stuff so that we can merge the other stuff? I have some shapes of ideas for radio button groups [though they are a bit messy].

BryceStevenWilley replied:
Discussion about this ~35 minutes in Deep dive recording

BryceStevenWilley committed:
split type != radio logic from custom datatype

Co-authored-by: plocket <52798256+plocket@users.noreply.github.com>

plocket committed:
Return default input type of 'text' when setting input see

comment at https://github.com/SuffolkLITLab/ALKiln/pull/633/files/a2db94dcc7e394497ec3c1b7d35a7f540126d0ff#r1134716311
See mdn docs at https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#input_types

plocket committed:
Remove radio button changes (#806)

* Fix errors in random tests

* Bump minor v to fix artifacts not being found

* Remove radio button changes

---------

Co-authored-by: Bryce Willey <Bryce.Steven.Willey@gmail.com>

plocket commented on lib/scope.js:
```diff
@@ -2231,9 +2240,14 @@ module.exports = {
     }  // ends act depending values
   },  // Ends scope.getThereIsAnotherValue()
 
+  is_on_interview_still: async function ( scope ) {
```
Change to `is_on_server_still`

plocket commented on lib/scope.js:
```diff
@@ -2737,7 +2751,25 @@ module.exports = {
+        field = faker.helpers.arrayElement( fields );
+      }
+      if (field == null) {
+        // Just skip it? Shouldn't happen but has before
```
Add a debug log here that'll give us more info in the future.

plocket commented on lib/scope.js:
```diff
@@ -2737,7 +2751,25 @@ module.exports = {
+      });
+      let field = null;
+      if (back_btn != null && Math.random() > 0.9) {
+        // 10% chance of clicking the back button
```
Explain why we want to check back button presses

plocket committed:
Add debug log, tweak var name, comment

plocket commented on lib/scope.js:
```diff
@@ -1401,8 +1401,7 @@ module.exports = {
       // interact with that label instead.
-      // `elem` has never been null from what we've seen
-      if ( elem.nextSibling && elem.nextSibling.tag === `label` ) {
+      if ( elem && elem.nextSibling && elem.nextSibling.tag === `label` ) {
```
Log non-existent elems for debugging: Not sure how to add a debug log outside of the DOM here, but we should log an element situation that doesn't exist and maybe have some kind of response to that. Or we shouldn't send selectors for elements that don't exist. I'm not sure how that's even happening. Maybe we can send in `reports` to write in or some such.

Same below in `get_custom_datatype()` and `enter_answer.input()`.

plocket replied:
Made https://github.com/SuffolkLITLab/ALKiln/issues/808

plocket committed:
Update changelog

plocket reviewed (approved):
Some great fixes! I know we won't get another chance to review, so I'm calling it.
